File: clients/matlab/simpleTest.m
% Make sure to have the add-on "ZMQ remote API"
% running in CoppeliaSim
%

fprintf('Program started\n')

client = RemoteAPIClient();
sim = client.getObject('sim');

% When simulation is not running, ZMQ message handling could be a bit
% slow, since the idle loop runs at 8 Hz by default. So let's make
% sure that the idle loop runs at full speed for this program:
defaultIdleFps = sim.getInt32Param(sim.intparam_idle_fps);
sim.setInt32Param(sim.intparam_idle_fps, 0);

% Create a few dummies and set their positions:
n = 50;
handles = zeros(n, 1);
for i=1:n
    handles(i) = sim.createDummy(0.01, zeros(12, 1));
    sim.setObjectPosition(handles(i), -1, [0.01 * i, 0.01 * i, 0.01 * i]);
end

% Run a simulation in asynchronous mode:
sim.startSimulation();

while 1
    t = sim.getSimulationTime();
    if t >= 3; break; end
    s = sprintf('Simulation time: %.2f [s] (simulation running asynchronously to client, i.e. non-stepped)', t);
    fprintf('%s\n', s);
    sim.addLog(sim.verbosity_scriptinfos, s);
end
sim.stopSimulation();
% If you need to make sure we really stopped:
while sim.getSimulationState() ~= sim.simulation_stopped
    pause(0.1);
end

% Run a simulation in stepping mode:
client.setStepping(true);
sim.startSimulation();
while 1
    t = sim.getSimulationTime();
    if t >= 3; break; end
    s = sprintf('Simulation time: %.2f [s] (simulation running synchronously to client, i.e. stepped)', t);
    fprintf('%s\n', s);
    sim.addLog(sim.verbosity_scriptinfos, s)
    client.step();  % triggers next simulation step
end
sim.stopSimulation();

% Remove the dummies created earlier:
for i=1:n
    sim.removeObject(handles(i));
end

% Restore the original idle loop frequency:
sim.setInt32Param(sim.intparam_idle_fps, defaultIdleFps);

fprintf('Program ended\n');

File: clients/python/simpleTest.py
# Make sure to have the add-on "ZMQ remote API"
# running in CoppeliaSim
#
# All CoppeliaSim commands will run in blocking mode (block
# until a reply from CoppeliaSim is received). For a non-
# blocking example, see simpleTest-nonBlocking.py

import time

from coppeliasim_zmqremoteapi_client import RemoteAPIClient

def myFunc(input1, input2):
    print('Hello', input1, input2)
    return 21

print('Program started')

client = RemoteAPIClient()
sim = client.require('sim')

# Create a few dummies and set their positions:
handles = [sim.createDummy(0.01, 12 * [0]) for _ in range(50)]
for i, h in enumerate(handles):
    sim.setObjectPosition(h, -1, [0.01 * i, 0.01 * i, 0.01 * i])

# Run a simulation in asynchronous mode:
sim.startSimulation()
while (t := sim.getSimulationTime()) < 3:
    s = f'Simulation time: {t:.2f} [s] (simulation running asynchronously '\
        'to client, i.e. non-stepping)'
    print(s)
    sim.addLog(sim.verbosity_scriptinfos, s)
    # sim.testCB(21,myFunc,42) # see below. sim.testCB is calling back above "myFunc"

# e.g. calling a script object function (make sure the script is running!):
'''    
script = sim.getObject('/path/to/scriptObject')
reply = sim.callScriptFunction('functionName', script, 'Hello', 'Paul', 21)
'''
#or
'''
script = sim.getObject('/path/to/scriptObject')
funcs = client.getScriptFunctions(script)
reply = funcs.functionName('Hello', 'Paul', 21)
'''    
    
sim.stopSimulation()
# If you need to make sure we really stopped:
while sim.getSimulationState() != sim.simulation_stopped:
    time.sleep(0.1)

# Run a simulation in stepping mode:
sim.setStepping(True)
sim.startSimulation()
while (t := sim.getSimulationTime()) < 3:
    s = f'Simulation time: {t:.2f} [s] (simulation running synchronously '\
        'to client, i.e. stepping)'
    print(s)
    sim.addLog(sim.verbosity_scriptinfos, s)
    sim.step()  # triggers next simulation step
sim.stopSimulation()

# Remove the dummies created earlier:
for h in handles:
    sim.removeObject(h)

print('Program ended')

''' test callback on CoppeliaSim side:
int ret = sim.testCB(int a, func cb, int b)
function sim.testCB(a, cb, b)
    for i = 1, 99, 1 do
        cb(a, b)
    end
    return cb(a, b)
end
'''
